### ISSUE TITLE
Question about the Loss Function

### ISSUE THREAD

**Issue by namin-an** [ExplainableML/sae-for-vlm#2]
Thank you for such an inspiring work!
I just have a question regarding the loss function of Matryoshka SAE.
From equation (4) in your paper (p.4), it seems like you have `|| v - W_dec ReLU(v) ||_2`.

However, in your code (https://github.com/ExplainableML/sae-for-vlm/blob/69e211dc4d0a31cfb3b3f6f682f78a1890677d5f/dictionary_learning/trainers/matroyshka_batch_top_k.py#L288), 
it seems like you input `residual_BD.float()` instead of `x`, leading to `|| (v - v_hat) - W_dec ReLU (post_relu_feat_acts_BF) ||_2`,
where `post_relu_feat_acts_BF = nn.functional.relu((x - self.b_dec) @ self.W_enc + self.b_enc)`.

Could you please explain this a bit further?
Thank you!

**Comment by mateuszpach** [comment]
Thank you for your interest in our work!  

The loss function referred to in the paper is the **L2 reconstruction loss**, implemented here:  
https://github.com/ExplainableML/sae-for-vlm/blob/69e211dc4d0a31cfb3b3f6f682f78a1890677d5f/dictionary_learning/trainers/matroyshka_batch_top_k.py#L284  

In contrast, the code you pointed out comes from this function:  
https://github.com/ExplainableML/sae-for-vlm/blob/69e211dc4d0a31cfb3b3f6f682f78a1890677d5f/dictionary_learning/trainers/matroyshka_batch_top_k.py#L214  

That part implements an **auxiliary loss** designed to prevent dead latents. For brevity, we did not describe this in the paper, but you can find more details in [this work (see Section A.2)](https://arxiv.org/abs/2406.04093v1).

**Comment by namin-an** [comment]
Thank you for your comment! I will close this issue!
